lib/iris/tests/test_concatenate.py

```diff
@@ -335,113 +337,117 @@ def test_order_difference(self):
         y = (0, 2)
-        cube = _make_cube((0, 2), y, 1, mask=True)
-        cube.fill_value = 10
-        cubes.append(cube)
-        cube = _make_cube((2, 4), y, 1, mask=True)
-        cube.fill_value = 20
+        cubes.append(_make_cube((0, 2), y, 1))
+        cube = _make_cube((2, 4), y, 2, mask=True)
         cubes.append(cube)
-        emsg = 'Fill values differ'
-        with self.assertRaisesRegexp(iris.exceptions.ConcatenateError, emsg):
-            cubes.concatenate_cube()
-
+        result = concatenate(cubes)
+        self.assertEqual(len(result), 1)
+        self.assertIsNone(result[0].fill_value)
+        np_fill_value = ma.masked_array(0, dtype=result[0].dtype).fill_value
+        self.assertEqual(result[0].data.fill_value, np_fill_value)
 
-class Test2D(tests.IrisTest):
-    def test_masked_and_unmasked(self):
+    def test_masked_and_unmasked__fill_value_equal(self):
         cubes = []
         y = (0, 2)
         fill_value = 1234
         cube = _make_cube((0, 2), y, 1, mask=True,
                           fill_value=fill_value)
         cubes.append(cube)
-        cubes.append(_make_cube((2, 4), y, 2))
+        cubes.append(_make_cube((2, 4), y, 2,
+                                fill_value=fill_value))
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
         self.assertEqual(result[0].fill_value, fill_value)
         self.assertEqual(result[0].data.fill_value, fill_value)
 
-    def test_masked_and_unmasked_int16(self):
+    def test_masked_and_unmasked_int16__fill_value_equal(self):
         cubes = []
         y = (0, 2)
         dtype = np.dtype('int16')
         fill_value = 1234
         cube = _make_cube((0, 2), y, 1, dtype=dtype, mask=True,
                           fill_value=fill_value)
         cubes.append(cube)
-        cubes.append(_make_cube((2, 4), y, 2, dtype=dtype))
+        cubes.append(_make_cube((2, 4), y, 2, dtype=dtype,
+                                fill_value=fill_value))
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
         self.assertEqual(result[0].fill_value, fill_value)
         self.assertEqual(result[0].data.fill_value, fill_value)
 
-    def test_unmasked_and_masked(self):
+    def test_unmasked_and_masked__default_fill_value_from_None(self):
         cubes = []
         y = (0, 2)
         cubes.append(_make_cube((0, 2), y, 1))
-        fill_value = 1234
         cube = _make_cube((2, 4), y, 2, mask=True,
-                          fill_value=fill_value)
+                          fill_value=1234)
         cubes.append(cube)
         result = concatenate(cubes)
         self.assertEqual(len(result), 1)
-        self.assertEqual(result[0].fill_value, fill_value)
-        self.assertEqual(result[0].data.fill_value, fill_value)
+        self.assertIsNone(result[0].fill_value)
+        np_fill_value = ma.masked_array(0, dtype=result[0].dtype).fill_value
+        self.assertEqual(result[0].data.fill_value, np_fill_value)
```

Comment:
I'm wondering if we need this check. I think perhaps it should be part of the contract that if the input fill values don't match then the fill value you get out on the array shouldn't be relied upon to have any particular value. That said, we are making an effort to make sure the correct fill value is placed on the data array when it _is_ valid, so I'm not sure.